src/chrome/offscreen/wallet-connection/wallet-connection-client.ts

```diff
@@ -0,0 +1,160 @@
+import { Queue } from 'queues/queue'
+import { AppLogger, logger as utilsLogger } from 'utils/logger'
+import { Worker } from 'queues/worker'
+import { MessageClient } from 'chrome/messages/message-client'
+import { WalletConnectionMessageHandler } from './message-handler'
+import { Message } from 'chrome/messages/_types'
+import { Subscription } from 'rxjs'
+import { SyncClient } from './sync-client'
+
+export type WalletConnectionClient = ReturnType<typeof WalletConnectionClient>
+
+export const WalletConnectionClient = ({
+  messagesRouter,
+  connectionPassword,
+  logger = utilsLogger,
+  syncClient,
+  connectorClient,
+}: {
+  connectionPassword: string
+  messagesRouter: MessagesRouter
+  logger: AppLogger
+  syncClient: SyncClient
+  connectorClient: ConnectorClient
+}) => {
+  connectorClient.setConnectionPassword(Buffer.from(connectionPassword, 'hex'))
+
+  const dAppRequestQueue = Queue<WalletInteractionWithOrigin>({
+    key: 'dAppRequestQueue',
+    logger,
+    paused: true,
+    worker: Worker((job) =>
+      connectorClient
+        .sendMessage(job.data, { timeout: config.webRTC.confirmationTimeout })
+        .map(() =>
+          messagesRouter
+            .getByInteractionId(job.data.interactionId)
+            .andThen((metadata) =>
+              messageClient.sendMessageAndWaitForConfirmation(
+                createMessage.sendMessageEventToDapp(
+                  'offScreen',
+                  'receivedByWallet',
+                  { interactionId: job.data.interactionId, metadata },
+                ),
+                metadata.tabId,
+              ),
+            ),
+        )
+        .map((result) => {
+          syncClient.addConfirmedInteractionId(job.data.interactionId)
+          return result
+        })
+        .mapErr((error) => {
+          const retryIfNotConnected = error.reason === 'notConnected'
+          const retryIfMessageFailed = job.numberOfRetries < 3
+
+          return {
+            ...error,
+            shouldRetry: retryIfNotConnected || retryIfMessageFailed,
+          }
+        }),
+    ),
+  })
+
+  const ledgerToWalletQueue = Queue<LedgerResponse>({
+    key: 'ledgerToWallet',
+    logger,
+    worker: Worker((job) =>
+      connectorClient.sendMessage(job.data, {
+        timeout: config.webRTC.confirmationTimeout,
+      }),
+    ),
+  })
+
+  const incomingWalletMessageQueue = Queue<Record<string, any>>({
+    key: 'incomingWalletMessageQueue',
+    logger,
+    paused: false,
+    worker: Worker((job) =>
+      messageClient
+        .handleMessage(createMessage.incomingWalletMessage('wallet', job.data))
+        .mapErr((err) => ({
+          ...err,
+          shouldRetry: false,
+        })),
+    ),
+  })
+
+  const subscription = new Subscription()
+
+  subscription.add(
+    syncClient.dappRequestConfirmedByWallet$.subscribe((interactionId) => {
+      dAppRequestQueue.cancel(interactionId)
```

Comment:
Should only cancel jobs that are pending.

Reply:
Looks like the logic to only cancel pending jobs is already inside the queue

Reply:
Ok great!